File: src/test/java/uk/gov/justice/digital/hmpps/riskprofiler/services/LifeDecisionTreeServiceTest.java
package uk.gov.justice.digital.hmpps.riskprofiler.services;

import org.assertj.core.api.Assertions;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;
import uk.gov.justice.digital.hmpps.riskprofiler.model.OffenderBooking;
import uk.gov.justice.digital.hmpps.riskprofiler.model.OffenderSentenceTerms;

import java.util.List;

import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class LifeDecisionTreeServiceTest {

    private static final String OFFENDER_1 = "A1234AB";
    private static final Long BOOKING_1 = -1L;
    private LifeDecisionTreeService service;

    private final OffenderSentenceTerms lifeSentence = OffenderSentenceTerms.builder()
            .bookingId(BOOKING_1)
            .lifeSentence(true)
            .build();
    private final OffenderSentenceTerms nonLifeSentence = OffenderSentenceTerms.builder()
            .bookingId(BOOKING_1)
            .lifeSentence(false)
            .build();
    private final OffenderBooking lifeCode = OffenderBooking.builder()
            .bookingId(BOOKING_1)
            .imprisonmentStatus("CFLIFE")
            .build();
    private final OffenderBooking nonLifeCode = OffenderBooking.builder()
            .bookingId(BOOKING_1)
            .imprisonmentStatus("OTHER")
            .build();

    @Mock
    private NomisService nomisService;

    @Before
    public void setup() {
        service = new LifeDecisionTreeService(nomisService);
    }

    @Test
    public void testWhenLifeFlagTrue() {
        when(nomisService.getBooking(OFFENDER_1)).thenReturn(BOOKING_1);
        when(nomisService.getSentencesForOffender(BOOKING_1)).thenReturn(List.of(nonLifeSentence, lifeSentence));

        final var profile = service.getLifeProfile(OFFENDER_1);

        Assertions.assertThat(profile.getProvisionalCategorisation()).isEqualTo("B");
        Assertions.assertThat(profile.getNomsId()).isEqualTo(OFFENDER_1);
        Assertions.assertThat(profile.isLife()).isTrue();
    }

    @Test
    public void testWhenLifeCode() {
        when(nomisService.getBooking(OFFENDER_1)).thenReturn(BOOKING_1);
        when(nomisService.getSentencesForOffender(BOOKING_1)).thenReturn(List.of(nonLifeSentence));
        when(nomisService.getBookingDetails(BOOKING_1)).thenReturn(List.of(nonLifeCode, lifeCode));

        final var profile = service.getLifeProfile(OFFENDER_1);

        Assertions.assertThat(profile.getProvisionalCategorisation()).isEqualTo("B");
        Assertions.assertThat(profile.getNomsId()).isEqualTo(OFFENDER_1);
        Assertions.assertThat(profile.isLife()).isTrue();
    }

    @Test
    public void testWhenMurder() {
        when(nomisService.getBooking(OFFENDER_1)).thenReturn(BOOKING_1);
        when(nomisService.getSentencesForOffender(BOOKING_1)).thenReturn(List.of(nonLifeSentence));
        when(nomisService.getBookingDetails(BOOKING_1)).thenReturn(List.of(nonLifeCode));
        when(nomisService.getMainOffences(BOOKING_1)).thenReturn(List.of("Murder etc."));

        final var profile = service.getLifeProfile(OFFENDER_1);

        Assertions.assertThat(profile.getProvisionalCategorisation()).isEqualTo("B");
        Assertions.assertThat(profile.getNomsId()).isEqualTo(OFFENDER_1);
        Assertions.assertThat(profile.isLife()).isTrue();
    }

    @Test
    public void testWhenNotLife() {
        when(nomisService.getBooking(OFFENDER_1)).thenReturn(BOOKING_1);
        when(nomisService.getSentencesForOffender(BOOKING_1)).thenReturn(List.of(nonLifeSentence));
        when(nomisService.getBookingDetails(BOOKING_1)).thenReturn(List.of(nonLifeCode, nonLifeCode));
        when(nomisService.getMainOffences(BOOKING_1)).thenReturn(List.of("Trivial etc.", "another"));

        final var profile = service.getLifeProfile(OFFENDER_1);

        Assertions.assertThat(profile.getProvisionalCategorisation()).isEqualTo("C");
        Assertions.assertThat(profile.getNomsId()).isEqualTo(OFFENDER_1);
        Assertions.assertThat(profile.isLife()).isFalse();
    }
}

File: src/test/kotlin/uk/gov/justice/digital/hmpps/riskprofiler/services/LifeDecisionTreeServiceTest.kt
package uk.gov.justice.digital.hmpps.riskprofiler.services

import org.assertj.core.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.jupiter.MockitoExtension
import uk.gov.justice.digital.hmpps.riskprofiler.model.OffenderBooking
import uk.gov.justice.digital.hmpps.riskprofiler.model.OffenderSentenceTerms

@ExtendWith(MockitoExtension::class)
class LifeDecisionTreeServiceTest {
  private lateinit var service: LifeDecisionTreeService
  private val lifeSentence = OffenderSentenceTerms(BOOKING_1, true)
  private val nonLifeSentence = OffenderSentenceTerms(BOOKING_1, false)
  private val lifeCode = OffenderBooking(BOOKING_1, null, null, "CFLIFE")
  private val nonLifeCode = OffenderBooking(BOOKING_1, null, null, "OTHER")

  @Mock
  private val nomisService: NomisService? = null
  @BeforeEach
  fun setup() {
    service = LifeDecisionTreeService(nomisService!!)
  }

  @Test
  fun testWhenLifeFlagTrue() {
    Mockito.`when`(nomisService!!.getBooking(OFFENDER_1)).thenReturn(BOOKING_1)
    Mockito.`when`(nomisService.getSentencesForOffender(BOOKING_1)).thenReturn(listOf(nonLifeSentence, lifeSentence))
    val (nomsId, provisionalCategorisation, life) = service.getLifeProfile(OFFENDER_1)
    Assertions.assertThat(provisionalCategorisation).isEqualTo("B")
    Assertions.assertThat(nomsId).isEqualTo(OFFENDER_1)
    Assertions.assertThat(life).isTrue
  }

  @Test
  fun testWhenLifeCode() {
    Mockito.`when`(nomisService!!.getBooking(OFFENDER_1)).thenReturn(BOOKING_1)
    Mockito.`when`(nomisService.getSentencesForOffender(BOOKING_1)).thenReturn(listOf(nonLifeSentence))
    Mockito.`when`(nomisService.getBookingDetails(BOOKING_1)).thenReturn(listOf(nonLifeCode, lifeCode))
    val (nomsId, provisionalCategorisation, life) = service.getLifeProfile(OFFENDER_1)
    Assertions.assertThat(provisionalCategorisation).isEqualTo("B")
    Assertions.assertThat(nomsId).isEqualTo(OFFENDER_1)
    Assertions.assertThat(life).isTrue
  }

  @Test
  fun testWhenMurder() {
    Mockito.`when`(nomisService!!.getBooking(OFFENDER_1)).thenReturn(BOOKING_1)
    Mockito.`when`(nomisService.getSentencesForOffender(BOOKING_1)).thenReturn(listOf(nonLifeSentence))
    Mockito.`when`(nomisService.getBookingDetails(BOOKING_1)).thenReturn(listOf(nonLifeCode))
    Mockito.`when`(nomisService.getMainOffences(BOOKING_1)).thenReturn(listOf("Murder etc."))
    val (nomsId, provisionalCategorisation, life) = service.getLifeProfile(OFFENDER_1)
    Assertions.assertThat(provisionalCategorisation).isEqualTo("B")
    Assertions.assertThat(nomsId).isEqualTo(OFFENDER_1)
    Assertions.assertThat(life).isTrue
  }

  @Test
  fun testWhenNotLife() {
    Mockito.`when`(nomisService!!.getBooking(OFFENDER_1)).thenReturn(BOOKING_1)
    Mockito.`when`(nomisService.getSentencesForOffender(BOOKING_1)).thenReturn(listOf(nonLifeSentence))
    Mockito.`when`(nomisService.getBookingDetails(BOOKING_1)).thenReturn(listOf(nonLifeCode, nonLifeCode))
    Mockito.`when`(nomisService.getMainOffences(BOOKING_1)).thenReturn(listOf("Trivial etc.", "another"))
    val (nomsId, provisionalCategorisation, life) = service.getLifeProfile(OFFENDER_1)
    Assertions.assertThat(provisionalCategorisation).isEqualTo("C")
    Assertions.assertThat(nomsId).isEqualTo(OFFENDER_1)
    Assertions.assertThat(life).isFalse
  }

  companion object {
    private const val OFFENDER_1 = "A1234AB"
    private const val BOOKING_1 = -1L
  }
}
